Refactor scale-down for better integration with drainability rules

artemvmin opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This is one of several CLs to move all drain conditions to drainability rules. Once complete, clients implementing custom drainability rules will have full control over the scale-down of nodes.

Notable changes:

- Split out drainability rules into separate packages. `simulation/drainability:Rule.Drainable()` function now takes a `DrainContext`. This function can assume that `DrainContext` is not nil.
- Refactor NodeDeleteOptions for reuse in the drainability package. `simulation:NodeDeleteOptions` has been split into two structs: `simulation/options:NodeDeleteOptions` and `drainability/rules:Rules`. Consumers of this struct have been updated accordingly.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
None
```

x13n commented:
/assign

x13n commented:
/cc @olagacek

x13n commented on cluster-autoscaler/main.go:
```diff
@@ -469,7 +469,7 @@ func buildAutoscaler(debuggingSnapshotter debuggingsnapshot.DebuggingSnapshotter
 		sdCandidatesSorting := previouscandidates.NewPreviousCandidates()
 		scaleDownCandidatesComparers = []scaledowncandidates.CandidatesComparer{
-			emptycandidates.NewEmptySortingProcessor(emptycandidates.NewNodeInfoGetter(opts.ClusterSnapshot), deleteOptions),
+			emptycandidates.NewEmptySortingProcessor(emptycandidates.NewNodeInfoGetter(opts.ClusterSnapshot), deleteOptions, rules.Rules{}),
```
This should use default, not empty rule set.

artemvmin replied:
Good catch. Also updated other references to NewNodeDeleteOptions to use default rules.

artemvmin commented on cluster-autoscaler/processors/scaledowncandidates/emptycandidates/empty_candidates_sorting_test.go:
```diff
@@ -62,13 +63,16 @@ func TestScaleDownEarlierThan(t *testing.T) {
+	p := EmptySorting{
+		nodeInfoGetter:    &niGetter,
+		deleteOptions:     deleteOptions,
+		drainabilityRules: rules.Rules{},
```
Thanks for pointing this out. I had all references nil initially, but changed them to empty set for nil safety. I forgot that this changes the defaulting behavior.

Updated all tests (except the part of static_autoscaler_test.go which uses NewNodeDeleteOptions() explicitly.

artemvmin commented on cluster-autoscaler/simulator/drain.go:
```diff
@@ -86,12 +60,19 @@ func GetPodsToMove(nodeInfo *schedulerframework.NodeInfo, deleteOptions NodeDele
 		case drainability.BlockDrain:
-			blockingPod = &drain.BlockingPod{pod, d.BlockingReason}
-			err = d.Error
+			// TODO(reviewer note): What is the consequence of returning a pod when
```
Ack. Removed.

artemvmin commented on cluster-autoscaler/simulator/drain.go:
```diff
@@ -106,39 +87,6 @@ func GetPodsToMove(nodeInfo *schedulerframework.NodeInfo, deleteOptions NodeDele
 	if err != nil {
 		return pods, daemonSetPods, blockingPod, err
 	}
-	if pdbBlockingPod, err := checkPdbs(pods, pdbs); err != nil {
```
I became aware of this ordering issue early on and decided to turn this PR into the full story. I reordered the commits and updated the title. I'll ping for a follow-up review when the remainder is implemented.

x13n commented:
The current refactor looks good to me now. Are you sure you want to rewrite all the checks in one humongous PR? You may start getting into merge conflicts, so I'd suggest following up in a separate PR - WDYT?

Btw, in the release notes you're adding some actions required - while this is true for downstream/forked code, it may be confusing for OSS CA release notes (which is the intended audience for these). I don't think this change should have any user-visible changes.

artemvmin commented:
> The current refactor looks good to me now. Are you sure you want to rewrite all the checks in one humongous PR? You may start getting into merge conflicts, so I'd suggest following up in a separate PR - WDYT?

Sounds good. Thanks for the tip. I updated the PR title.

> Btw, in the release notes you're adding some actions required - while this is true for downstream/forked code, it may be confusing for OSS CA release notes (which is the intended audience for these). I don't think this change should have any user-visible changes.

That makes sense. Updated.

artemvmin commented on cluster-autoscaler/simulator/drainability/rules/pdb/pdb.go:
```diff
@@ -0,0 +1,54 @@
+	// require adding information to the DrainContext, such as the slice of pods
+	// and a flag to prevent duplicate checks.
+	for _, pdb := range drainCtx.Pdbs {
+		selector, err := metav1.LabelSelectorAsSelector(pdb.Spec.Selector)
```
I had a separate draft passing RemainingPdbTracker directly, but without using DrainCtx. The combination of these two ideas seems to be the sweet spot.

x13n replied:
Ah, good question. As far as I can tell though, you're using a new instance of the tracker in each such goroutine. This is perhaps suboptimal, but should be safe.

x13n commented on cluster-autoscaler/simulator/drain.go:
```diff
@@ -106,7 +90,7 @@ func GetPodsToMove(nodeInfo *schedulerframework.NodeInfo, deleteOptions NodeDele
 		return pods, daemonSetPods, blockingPod, err
 	}
-	if pdbBlockingPod, err := checkPdbs(pods, pdbs); err != nil {
+	if pdbBlockingPod, err := checkPdbs(pods, remainingPdbTracker.GetPdbs()); err != nil {
```
It seems you could already delete `checkPdbs` function now and just use `RemainingPdbTracker.CanRemovePods()`.

artemvmin replied:
Done. I wasn't sure what the "legacy scale-down" was referring to in the comment below so I left it alone. The logic looks identical if the parallel return value is ignored.

x13n commented:
/lgtm
/approve
/hold

I only really have one minor comment, but it doesn't block this PR, so feel free to cancel the hold if you don't want to address it now.

artemvmin commented:
Comments addressed.

/unhold

x13n commented:
Thanks!

/lgtm

x13n commented:
/lgtm
